orc8r/cloud/go/services/certifier/servicers/certifier.go

```diff
@@ -712,60 +718,166 @@ func (srv *CertifierServer) getPolicyDecisionFromTokenMany(tokens *certprotos.To
+	reqType := constants.Path
+	switch req.ResourceId.(type) {
+	case *certprotos.Request_NetworkId:
+		reqType = constants.NetworkID
+	case *certprotos.Request_TenantId:
+		reqType = constants.TenantID
+	}
+	return reqType
+}
+
+func getPolicyType(policy *certprotos.Policy) constants.ResourceType {
+	policyType := constants.Path
+	switch policy.Resource.(type) {
+	case *certprotos.Policy_Network:
+		policyType = constants.NetworkID
+	case *certprotos.Policy_Tenant:
+		policyType = constants.TenantID
+	}
+	return policyType
+}
+
+// doResourceTypesMatch checks if the policy type applies to the requested type
+// for network and tenant resource types.
+// Every request has a path but not necessarily network/tenant ids, so path policy
+// types are always applied to every request, and they are exempted from being matched by type.
+// Returns true if the resource type is either network or tenant and the resource types match
+// Returns false otherwise
+func doResourceTypesMatch(req *certprotos.Request, policy *certprotos.Policy) bool {
+	reqType := getRequestType(req)
+	policyType := getPolicyType(policy)
+	return req.GetResourceId() == nil || policyType == constants.Path || reqType == policyType
+}
+
+// isActionAuthorized checks if the requested action to read/write is authorized by the policy
+// Returns the effect if the policy allows/denies write access regardless of requested action,
+// or if policy allows/denies read access and the requested action is read
+// Otherwise, returns an unknown effect, which implies that the policy does not apply to this requested resource
+func isActionAuthorized(req *certprotos.Request, policy *certprotos.Policy) certprotos.Effect {
 	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
+	if !doResourceTypesMatch(req, policy) {
 		return certprotos.Effect_UNKNOWN
 	}
-	reqPath := reqResource.Resource
-	policyPath := policyResource.Resource
-	if ok, _ := doublestar.Match(policyPath, reqPath); ok {
-		return policyResource.Effect
+	if policy.Action == certprotos.Action_WRITE {
+		return policy.Effect
+	}
+	if policy.Action == certprotos.Action_READ && req.Action == certprotos.Action_READ {
+		return policy.Effect
 	}
 	return certprotos.Effect_UNKNOWN
 }
 
-// checkAction checks if the requested action is authorized by the policy
-func checkAction(reqResource *certprotos.Resource, policyResource *certprotos.Resource) certprotos.Effect {
+// isResourceAuthorized checks if the requested resource (either a path, network, or tenant) is authorized by the policy
+// Returns the effect if the policy's path matches the requested path and if the policy's tenant/network ID matches that
+// of the requested when applicable
+// Otherwise, returns an unknown effect, which implies that the policy does not apply to this requested resource
+func isResourceAuthorized(req *certprotos.Request, policy *certprotos.Policy) certprotos.Effect {
```

Comment:
Unless I'm mistaken, only 1 of these 3 blocks (path auth, network auth, tenant auth) will occur for any call to this fn. If that's an accurate description, then let's make that clear in the function body by using a switch statement or early returns